libs/arcade-core/arcade_core/usage/identity.py

```diff
@@ -124,34 +124,63 @@ def get_distinct_id(self) -> str:
         return str(data[KEY_ANON_ID])
 
     def get_principal_id(self) -> str | None:
-        """Fetch principal_id from Arcade Cloud API.
+        """Fetch principal_id (account_id) from Arcade Cloud API.
+
+        Prefers any linked principal already stored in usage.json. If not
+        found, attempts to fetch via OAuth access token and falls
+        back to legacy API key validation if present.
 
         Returns:
             str | None: Principal ID if authenticated and API call succeeds, None otherwise
         """
+        # Prefer already-linked principal_id in usage.json
+        data = self.load_or_create()
+        linked_principal_id = data.get(KEY_LINKED_PRINCIPAL_ID)
+        if linked_principal_id:
+            return str(linked_principal_id)
+
         if not os.path.exists(CREDENTIALS_FILE_PATH):
             return None
 
         try:
             with open(CREDENTIALS_FILE_PATH) as f:
-                config = yaml.safe_load(f)
-
-            cloud_config = config.get("cloud", {})
-            api_key = cloud_config.get("api", {}).get("key")
-
-            if not api_key:
-                return None
-
-            response = httpx.get(
-                "https://cloud.arcade.dev/api/v1/auth/validate",
-                headers={"accept": "application/json", "Authorization": f"Bearer {api_key}"},
-                timeout=TIMEOUT_ARCADE_API,
+                config = yaml.safe_load(f) or {}
+
+            cloud_config = config.get("cloud", {}) if isinstance(config, dict) else {}
+
+            # OAuth credentials: use access_token to call /whoami
+            auth_config = cloud_config.get("auth", {}) if isinstance(cloud_config, dict) else {}
+            access_token = auth_config.get("access_token")
+            if access_token:
+                response = httpx.get(
+                    "https://api.arcade.dev/api/v1/auth/whoami",
```

Comment:
I think this needs to be `https://cloud...`, not `https://api...` otherwise we'll get a 404 causing users to remain anonymous. 

Also, should we pass the coordinator_url to this function so that we can switch between prod & bosslevel?